course_UvA-DL/activation-functions/Activation_Functions.py

```diff
@@ -0,0 +1,824 @@
+
+plt.figure(figsize=(8, 8))
+plt.title("FashionMNIST examples")
+plt.imshow(img_grid)
+plt.axis('off')
+plt.show()
+plt.close()
+
+# %% [markdown]
+# ### Visualizing the gradient flow after initialization
+#
+# As mentioned previously, one important aspect of activation functions is how they propagate gradients through the network.
+# Imagine we have a very deep neural network with more than 50 layers.
+# The gradients for the input layer, i.e. the very first layer, have passed >50 times the activation function, but we still want them to be of a reasonable size.
+# If the gradient through the activation function is (in expectation) considerably smaller than 1, our gradients will vanish until they reach the input layer.
+# If the gradient through the activation function is larger than 1, the gradients exponentially increase and might explode.
+#
+# To get a feeling of how every activation function influences the
+# gradients, we can look at a freshly initialized network and measure the
+# gradients for each parameter for a batch of 256 images:
+
+
+# %%
+def visualize_gradients(net, color="C0"):
+    """
+    Inputs:
+        net - Object of class BaseNetwork
+        color - Color in which we want to visualize the histogram (for easier separation of activation functions)
+    """
+    net.eval()
+    small_loader = data.DataLoader(train_set, batch_size=256, shuffle=False)
+    imgs, labels = next(iter(small_loader))
+    imgs, labels = imgs.to(device), labels.to(device)
+
+    # Pass one batch through the network, and calculate the gradients for the weights
+    net.zero_grad()
+    preds = net(imgs)
+    loss = F.cross_entropy(preds, labels)
+    loss.backward()
+    # We limit our visualization to the weight parameters and exclude the bias to reduce the number of plots
+    grads = {
+        name: params.grad.data.view(-1).cpu().clone().numpy()
+        for name, params in net.named_parameters() if "weight" in name
+    }
+    net.zero_grad()
+
+    # Plotting
+    columns = len(grads)
+    fig, ax = plt.subplots(1, columns, figsize=(columns * 3.5, 2.5))
+    fig_index = 0
+    for key in grads:
+        key_ax = ax[fig_index % columns]
+        sns.histplot(data=grads[key], bins=30, ax=key_ax, color=color, kde=True)
+        key_ax.set_title(str(key))
+        key_ax.set_xlabel("Grad magnitude")
+        fig_index += 1
+    fig.suptitle(
+        f"Gradient magnitude distribution for activation function {net.config['act_fn']['name']}",
+        fontsize=14,
+        y=1.05
+    )
+    fig.subplots_adjust(wspace=0.45)
+    plt.show()
+    plt.close()
+
+
+# %%
+# Seaborn prints warnings if histogram has small values. We can ignore them for now
+warnings.filterwarnings('ignore')
+# Create a plot for every activation function
+for i, act_fn_name in enumerate(act_fn_by_name):
+    set_seed(
+        42
+    )  # Setting the seed ensures that we have the same weight initialization for each activation function
```

Comment:
```suggestion
    # Setting the seed ensures that we have the same weight initialization for each activation function
    set_seed(42)
```
any formatter does not like this long comments after, so maybe rather move these long comments above the line?